backend/infrahub/core/migrations/graph/m047_backfill_or_null_display_label.py

```diff
@@ -0,0 +1,645 @@
+            """ % {"branch_filter": branch_filter}
+        self.add_to_query(update_value_query)
+
+        if self.is_large_type_attribute:
+            set_value_query = """
+// ------------
+// only make updates if the existing value is not the same as the new value
+// ------------
+WITH attr, existing_av, value_id_pairs, $values_by_id[n.uuid] AS required_value
+WHERE existing_av.value <> required_value
+OR existing_av IS NULL
+WITH attr, value_id_pairs, required_value,
+    reduce(av_vertex_id = NULL, pair IN value_id_pairs |
+        CASE
+            WHEN av_vertex_id IS NOT NULL THEN av_vertex_id
+            WHEN pair[0] = required_value THEN pair[1]
+            ELSE av_vertex_id
+        END
+    ) AS av_vertex_id
+MATCH (av:AttributeValue)
+WHERE elementId(av) = av_vertex_id
+CREATE (attr)-[r:HAS_VALUE { branch: $branch, branch_level: $branch_level, status: "active", from: $at }]->(av)
+            """
+        else:
+            set_value_query = """
+// ------------
+// only make updates if the existing value is not the same as the new value
+// ------------
+WITH n, attr, existing_av, value_id_pairs, $values_by_id[n.uuid] AS required_value
+WHERE existing_av.value <> required_value
+OR existing_av IS NULL
+CALL (n, attr) {
+    MERGE (av:AttributeValue&AttributeValueIndexed {is_default: false, value: $values_by_id[n.uuid]} )
+    WITH av, attr
+    LIMIT 1
+    CREATE (attr)-[r:HAS_VALUE { branch: $branch, branch_level: $branch_level, status: "active", from: $at }]->(av)
+}
+            """
+        self.add_to_query(set_value_query)
+
+
+class GetNodesWithoutDisplayLabelQuery(Query):
+    """Get all active nodes that do not have a display_label attribute."""
+
+    name = "get_nodes_without_display_label"
+    type = QueryType.READ
+
+    def __init__(self, kinds_to_skip: list[str] | None = None, **kwargs: Any) -> None:
+        super().__init__(**kwargs)
+        self.kinds_to_skip = kinds_to_skip or []
+
+    async def query_init(self, db: InfrahubDatabase, **kwargs: dict[str, Any]) -> None:  # noqa: ARG002
+        self.params = {
+            "branch_names": [registry.default_branch, GLOBAL_BRANCH_NAME],
+            "kinds_to_skip": self.kinds_to_skip,
+            "attribute_name": "display_label",
+        }
+        query = """
+// ------------
+// Get all active nodes that don't have a display_label attribute
+// ------------
+MATCH (n:Node)-[e:IS_PART_OF]->(:Root)
+WHERE NOT n.kind IN $kinds_to_skip
+AND e.branch IN $branch_names
+AND e.status = "active"
+AND e.to IS NULL
+AND NOT exists((n)-[:IS_PART_OF {branch: e.branch, status: "deleted"}]->(:Root))
+WITH DISTINCT n, e AS is_part_of_e
+OPTIONAL MATCH (n)-[r:HAS_ATTRIBUTE]->(attr:Attribute {name: $attribute_name})
+WHERE r.branch IN $branch_names
+AND r.status = "active"
+AND r.to IS NULL
+WITH n, is_part_of_e, r AS has_attr_e
+WHERE is_part_of_e.status = "active" AND (has_attr_e IS NULL OR has_attr_e.status = "deleted")
+WITH n.uuid AS node_uuid
```

Comment:
- the `is_part_of_e` has already been filtered to only `status="active"` edges, so that can be removed
- `has_attr_e` will never be `status="deleted"` b/c it is filtered to `status="active"` in the first `WHERE` clause in the snippet below. I think you want to get the latest edge linking `n` to `:Attribute {name: $attribute_name}` and filter to only those rows where the latest `has_attr_e` is NULL or is deleted

I have not tested this, but I think this will do it

```suggestion
WITH DISTINCT n
CALL (n) {
    OPTIONAL MATCH (n)-[r:HAS_ATTRIBUTE]->(attr:Attribute {name: $attribute_name})
    WHERE r.branch IN $branch_names
    RETURN r AS has_attr_e
    ORDER BY r.from DESC, r.status ASC
    LIMIT 1
}
WITH n, has_attr_e
WHERE (has_attr_e IS NULL OR has_attr_e.status = "deleted")
WITH n.uuid AS node_uuid
```

also, not sure how many UUIDs this might be, but it could be necessary to paginate this if it could be more than roughly 10k